x-pack/solutions/security/plugins/security_solution/server/lib/siem_migrations/rules/task/retrievers/rule_migrations_retriever.ts

```diff
@@ -34,13 +35,17 @@ export class RuleMigrationsRetriever {
     this.prebuiltRules = new PrebuiltRulesRetriever(clients);
   }
 
+  private async populateIndices() {
+    await Promise.all([this.prebuiltRules.populateIndex(), this.integrations.populateIndex()]);
+  }
+
   public async initialize() {
-    await Promise.all([
-      this.resources.initialize(),
-      this.prebuiltRules.populateIndex(),
-      this.integrations.populateIndex(),
-    ]).catch((error) => {
-      throw new Error(`Failed to initialize RuleMigrationsRetriever: ${error}`);
-    });
+    // Run only one populateIndices promise at a time, if one is already running, wait for it to finish
+    if (RuleMigrationsRetriever.populatePromise === null) {
+      RuleMigrationsRetriever.populatePromise = this.populateIndices().finally(() => {
+        RuleMigrationsRetriever.populatePromise = null;
+      });
+    }
+    await Promise.all([RuleMigrationsRetriever.populatePromise, this.resources.initialize()]);
```

Comment:
I think i need to understand it better.

Let's say user has run 2 migrations sequentially and now first migrations has already got the handle of `populatePromise`. 

Consequently, when `2nd` migration calls `initialize` ( `initialize` by `1st` migration is also running ) it will simply skip the call to `populate` indices. Is this what we want?

Do we need to run the initialization of  `IntegrationRetriever` and `PrebuiltRulesRetriever` only once?

Reply:
@logeekal Yes, the initialization of these 2 retrievers is populating their corresponding ELSER indices. We don't need to do that per migration, if one population promise is going on, there's no need to try indexing the same documents again at the same time.

> Consequently, when 2nd migration calls initialize ( initialize by 1st migration is also running ) it will simply skip the call to populate indices. Is this what we want?

It will not simply skip it, it will _await_ the existing `populatePromise` from the first migration to finish before continuing, so no rule translation will start until the ELSER indices are populated.

Reply:
> if one population promise is going on, there's no need to try indexing the same documents again at the same time.

Okay understood. Thanks for explanation @semd. One follow up question.

 Let's say 2 migrations have run already and completed. Now i am running a third migration, then this `populateElserIndices` will run again. Why don't we run only if and if there is no successful migration existing.

Reply:
Got it. Thanks